### PR TITLE
[flutter_tool] Build a Fuchsia package

### PR DESCRIPTION
**Pull request by zanderso**
## Description

Adds a flutter tool build command for building a Fuchsia package.

## Related Issues

https://github.com/flutter/flutter/issues/31144

## Tests

I added the following tests:

flutter_tools/test/commands/build_fuchsia_test.dart

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

### CODE REVIEW

File: packages/flutter_tools/lib/src/build_info.dart [hunk 1]
```diff
@@ -382,7 +382,7 @@ String getIosBuildDirectory() {
   return fs.path.join(getBuildDirectory(), 'ios');
 }
 
-/// Returns the macOS build output directory.
+/// Returns the MacOS build output directory.
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/build_info.dart]: actually it is macOS

Comment (zanderso) [reply packages/flutter_tools/lib/src/build_info.dart]: Fixed

**Comment by zanderso** [comment]
Factored out invoking the kernel compiler and pm tool to add another test. PTAL.

File: packages/flutter_tools/lib/src/fuchsia/fuchsia_kernel_compiler.dart [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+      ];
+    } else {
+      throwToolExit('Expected build type to be debug, profile, or release');
+    }
+
+    flags += <String>[
+      '$multiRootScheme:///$target',
+    ];
+
+    final List<String> command = <String>[
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/fuchsia/fuchsia_kernel_compiler.dart]: The formatting here is a bit awkward, maybe:

```
final List<String> command = <String>[
  artifacts.getArtifactPath(Artifact.engineDartBinary),
  fuchsiaArtifacts.kernelCompiler.path,
  ..flags,
]
```

Or

```
final List<String> command = <String>[
  artifacts.getArtifactPath(Artifact.engineDartBinary),
  fuchsiaArtifacts.kernelCompiler.path,
]..addAll(flags);
```

File: packages/flutter_tools/lib/src/fuchsia/fuchsia_kernel_compiler.dart [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+class FuchsiaKernelCompiler {
+  /// Compiles the [fuchsiaProject] with entrypoint [target] to a collection of
+  /// .dilp files, and a manifest that refers to them.
+  Future<void> build({
+    @required FuchsiaProject fuchsiaProject,
+    @required String target, // E.g., lib/main.dart
+    BuildInfo buildInfo = BuildInfo.debug,
+  }) async {
+    const String multiRootScheme = 'main-root';
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/fuchsia/fuchsia_kernel_compiler.dart]: Will it be possible to make this configurable eventually? We also use a multi-root scheme to support the build_runner cache directory. Doesn't need to be handled in this PR.

File: packages/flutter_tools/lib/src/fuchsia/fuchsia_kernel_compiler.dart [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+import 'fuchsia_sdk.dart';
+
+/// The [FuchsiaKernelCompiler] instance.
+FuchsiaKernelCompiler get fuchsiaKernelCompiler =>
+    context.get<FuchsiaKernelCompiler>();
+
+class FuchsiaKernelCompiler {
+  /// Compiles the [fuchsiaProject] with entrypoint [target] to a collection of
+  /// .dilp files, and a manifest that refers to them.
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/fuchsia/fuchsia_kernel_compiler.dart]: What are the rules for how an app is split into .dilp files? package boundary or is it more ... arbitrary?

File: packages/flutter_tools/lib/src/fuchsia/fuchsia_pm.dart [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+// found in the LICENSE file.
+
+import '../base/context.dart';
+import '../base/io.dart';
+import '../base/process_manager.dart';
+import '../globals.dart';
+
+import 'fuchsia_sdk.dart';
+
+/// The [FuchsiaSdk] instance.
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/fuchsia/fuchsia_pm.dart]:
```suggestion
/// The [FuchsiaPM] instance.
```

File: packages/flutter_tools/lib/src/fuchsia/fuchsia_pm.dart [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+FuchsiaPM get fuchsiaPM => context.get<FuchsiaPM>();
+
+class FuchsiaPM {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/fuchsia/fuchsia_pm.dart]: Would love some high level documentation explaining what each command does (even if it mostly points to existing pm documentation)

**Review by jonahwilliams — APPROVED** [review]
Generally LGTM, with some comments on formatting and documentation.

File: packages/flutter_tools/lib/src/fuchsia/fuchsia_kernel_compiler.dart (hunk 2 above)
Comment (zanderso) [reply packages/flutter_tools/lib/src/fuchsia/fuchsia_kernel_compiler.dart]: Done

File: packages/flutter_tools/lib/src/fuchsia/fuchsia_kernel_compiler.dart (hunk 3 above)
Comment (zanderso) [reply packages/flutter_tools/lib/src/fuchsia/fuchsia_kernel_compiler.dart]: Added a TODO to pick up the filesystem roots and scheme from the BuildInfo.

File: packages/flutter_tools/lib/src/fuchsia/fuchsia_kernel_compiler.dart (hunk 4 above)
Comment (zanderso) [reply packages/flutter_tools/lib/src/fuchsia/fuchsia_kernel_compiler.dart]: Added a comment. (It is potentially arbitrary.)

File: packages/flutter_tools/lib/src/fuchsia/fuchsia_pm.dart (hunk 5 above)
Comment (zanderso) [reply packages/flutter_tools/lib/src/fuchsia/fuchsia_pm.dart]: Thanks!

File: packages/flutter_tools/lib/src/fuchsia/fuchsia_pm.dart (hunk 6 above)
Comment (zanderso) [reply packages/flutter_tools/lib/src/fuchsia/fuchsia_pm.dart]: Done.

**Pull request merged**